modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -488,6 +500,9 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
             self._log["epochs"][epoch]["OnBatchEnd"] = stopw.measurements.get("OnBatchEnd", 0)
 
             self._persist_pipeline_log()
+            if 0 < self.num_samples_to_pass <= self._num_samples:
+                self._info("reached the threshold of samples to pass; break out of epoch loop to stop training")
+                break
```

Comment:
Why do we need to check this twice?

Reply:
We have two loops here. One for each epoch and one for each batch in an epoch. When we reach the limit we need to break out of both loops.